### ISSUE TITLE
many posts go to rest.api.chatid.com

### ISSUE THREAD

**Issue by kamalmarhubi** [recursecenter/blaggregator#161]
chatid.com goes here: https://welcomecommerce.com/#consumer-intelligence

I've seen this in more than one baggregator URL from more than one author. Here's an example: https://blaggregator.recurse.com/post/n84rOc/view

From [the discussion thread](https://recurse.zulipchat.com/#narrow/stream/24328-blogging/subject/Some.20talks/near/117976130) it appears this was not always the case.

**Comment by kamalmarhubi** [comment]
Another example: https://blaggregator.recurse.com/post/p9dlHn/view
Discussion: https://recurse.zulipchat.com/#narrow/stream/24328-blogging/subject/Writing.20a.20postmortem.3A.20an.20interview.20exercise.20I.20really.C2.A0like/near/117885350

**Comment by kamalmarhubi** [comment]
And another, much more recent: https://blaggregator.recurse.com/post/Hc05vp/view
Discussion: https://recurse.zulipchat.com/#narrow/stream/24328-blogging/subject/Versioning.20and.20Release.20Process/near/123949656

**Comment by punchagan** [comment]
@kamalmarhubi Thanks for filing this issue. I've never seen this myself, and wasn't aware of this happening. I'll look into what's going on, here.

**Comment by punchagan** [comment]
The links you posted here, take me to the right posts for me.  I wonder if it's some DNS resolution thing, @kamalmarhubi

**Comment by punchagan** [comment]
I ran curl in a loop, and this is what I see!
![image](https://user-images.githubusercontent.com/315678/37814871-5cb85612-2e92-11e8-8597-9c93f389cf68.png)

The corresponding Heroku logs look like this
![image](https://user-images.githubusercontent.com/315678/37814856-4e65d152-2e92-11e8-8185-a98b25ae819f.png)


The size of the data being sent back in each response is the same, but some requests redirect me to some other locations.  The response content is also different.  

@davidbalbert would you have any ideas on what might be going on?

**Comment by punchagan** [comment]
@davidbalbert 
I added more logging, and did some more digging.  I sent a bunch of requests using `curl` and the requests which redirect me to the wrong server are actually not responses from the app.  Those requests don't seem to reach the app at all, which makes me think it is a DNS issue.  

Also, doing a `dig` on `blaggregator.recurse.com` shows me it is a `CNAME` for `proxy.recurse.com` and that has two `A` type ip mappings - `52.20.208.119` and `34.238.32.87`, whose order seems to keep changing randomly.  I wonder if one of the proxy servers is returning the incorrect addresses, and is the cause for this bug.

**Comment by davidbalbert** [comment]
Huh. I will try to look into this in the next few days.

**Comment by punchagan** [comment]
> Also, doing a dig on blaggregator.recurse.com shows me it is a CNAME for proxy.recurse.com and that has two A type ip mappings - 52.20.208.119 and 34.238.32.87, whose order seems to keep changing randomly. I wonder if one of the proxy servers is returning the incorrect addresses, and is the cause for this bug.

I tested this hypothesis out, and it is incorrect.  I get the incorrect redirect URL when DNS is resolved to any proxy.

**Comment by jvns** [comment]
looks like DNS issues have already been ruled out but in case it's interesting: here's a `curl` incantation that connects to https://blaggregator.recurse.com and forces it to resolve to a specific IP (so that no DNS servers are involved at any point)

```
curl https://blaggregator.recurse.com:443/ -I --resolve blaggregator.recurse.com:443:52.20.208.119 | grep Location
```

**Comment by punchagan** [comment]
> looks like DNS issues have already been ruled out but in case it's interesting

Thanks @jvns.  I just modified my `/etc/hosts` file, and it looks like it's not DNS related

**Comment by punchagan** [comment]
Also, here is a `curl -v` output where we get redirected to the wrong `Location`. 

```sh
$ curl -4 -v -i -s https://blaggregator.recurse.com:443/new/
*   Trying 52.20.208.119...
* TCP_NODELAY set
* Connected to blaggregator.recurse.com (52.20.208.119) port 443 (#0)
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: CN=recurse.com
*  start date: Mar  6 00:00:00 2018 GMT
*  expire date: Apr  6 12:00:00 2019 GMT
*  subjectAltName: host "blaggregator.recurse.com" matched cert's "*.recurse.com"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
> GET /new/ HTTP/1.1
> Host: blaggregator.recurse.com
> User-Agent: curl/7.55.1
> Accept: */*
> 
< HTTP/1.1 301 Moved Permanently
HTTP/1.1 301 Moved Permanently
< Content-Type: text/html
Content-Type: text/html
< Date: Fri, 23 Mar 2018 17:29:48 GMT
Date: Fri, 23 Mar 2018 17:29:48 GMT
< Location: https://rest.api.chatid.com/new/
Location: https://rest.api.chatid.com/new/
< Server: nginx/1.4.6 (Ubuntu)
Server: nginx/1.4.6 (Ubuntu)
< Strict-Transport-Security: max-age=63072000
Strict-Transport-Security: max-age=63072000
< Content-Length: 184
Content-Length: 184
< Connection: keep-alive
Connection: keep-alive

< 
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.4.4</center>
</body>
</html>
* Connection #0 to host blaggregator.recurse.com left intact
```

**Comment by davidbalbert** [comment]
When I get the correct response, it comes with the header "Via: 1.1 vegur."
Vegur <https://github.com/heroku/vegur> is Heroku's HTTP proxy. This makes
sense because blaggregator is hosted on Heroku.

When I get the wrong response, it doesn't include the "Via" header. This
makes me think it's likely that the bad requests are never reaching Heroku
and the problem is inside Proxy <https://github.com/recursecenter/proxy> (our
HTTP proxy). That said, perhaps there are other reasons that the Via header
might not get added.

Both responses include "Server: nginx/1.4.6 (Ubuntu)," but I wouldn't be
surprised if nginx on Proxy is including that header in both cases, but I'm
not sure.

I'm not yet looking into this in earnest. I just wanted to leave this note
here for other intrepid sleuths :).

**Comment by punchagan** [comment]
FWIW, running similar `curl` requests on `blaggregator.herokuapp.com` doesn't ever return the wrong response. So, it appears like the problem is before we hit Heroku.

**Comment by jasonaowen** [comment]
I was able to reproduce this on my app, [jasonaowen/faces](https://github.com/jasonaowen/recurse-faces/), as well as on the [niceties app](https://github.com/mjec/rc-niceties), all of which occasionally return ChatID-related pages, so it doesn't seem to be specific to blaggregator.

I've opened an issue for this at https://github.com/recursecenter/proxy/issues/3 .

**Comment by punchagan** [comment]
I've deployed a temporary fix to Blaggregator that uses the direct herokuapp URLs, instead of the recurse subdomain URLs. This should prevent the links posted by the bot from here on from having this problem.

**Comment by punchagan** [comment]
Closed in favor of recursecenter/proxy#3 .
